src/chipper/keyboard2.cljs

```diff
@@ -0,0 +1,168 @@
+;; Available macro types are
+;; :simple   (a sequence of position/frame/mode/attr)
+;; :advanced (a callable?)
+;; and macros mappings are defined as
+;; {:Trigger [:macro {:type type :value value}]}
+;; See edit mode :Backspace for an example
+
+(def measure-size 4)
+
+(def -motions
+  {:down  [ 1  0  0]
+   :up    [-1  0  0]
+   :left  [ 0  0 -1]
+   :right [ 0  0  1]})
+
+(def normal-keymap
+  {;; relative movement
+   ;; keeping :position for clarity. The representation for cursor position may
+   ;; change in the future, thus -motions
+   :KeyJ        [:position (:down  -motions)]
+   :KeyK        [:position (:up    -motions)]
+   :KeyH        [:position (:left  -motions)]
+   :KeyL        [:position (:right -motions)]
+
+   :KeyEnter    [:position (:down  -motions)]
+   :ArrowDown   [:position (:down  -motions)]
+   :ArrowUp     [:position (:up    -motions)]
+   :ArrowLeft   [:position (:left  -motions)]
+   :ArrowRight  [:position (:right -motions)]
+
+   :KeyW        [:position [ 4  0  0]]
+   :KeyB        [:position [-4  0  0]]
+
+   :ShiftBracketRight [:frame  1]
+   :BracketRight      [:frame -1]
+
+   ;; absolute movement
+   :KeyG        [:position [:first :first :first]]
+   :ShiftKeyG   [:position [:last  :last  :last]]
+   :Digit0      [:position [:first :first :first]]
+   :ShiftDigit4 [:position [:last  :last  :last]]
+
+   ;; mode change
+   :KeyI [:mode :edit]
+   :KeyV [:mode :visual]
+
+   ;; other
+   :KeyX        [:macro {:type :simple :value [[:attr nil] (:down-line -motions)]}]
+   ;; so these could be special or just, like... :tempo, and :octave
+   ;; I'm leaving them out for now
+   ;; :Period [:special [:tempo 1]]
+   ;; :ShiftPeriod [:special [:tempo -1]]
+   :Space [:special :play-pause]})
+
+(def edit-keymap
+  ;; hm, this isn't going to work quite like you want it to.
+  ;; this is fine for inputting notes...
+  {:KeyA [:attr :C]    :KeyW [:attr :C#]
+   :KeyS [:attr :D]    :KeyE [:attr :D#]
+   :KeyD [:attr :E]    :KeyR [:attr :F]
+   :KeyF [:attr :F#]   :KeyT [:attr :G]
+   :KeyG [:attr :G#]   :KeyY [:attr :A]
+   :KeyH [:attr :A#]   :KeyU [:attr :B]
+
+   ;; ... but when you get to the other attributes, what happens? They have
+   ;; values in [0x00, 0xff].
+   ;; You gotta either have another sort of translation layer where some of your
+   ;; note names become hex digits, or... something
+   ;; In fact, now that I think about it, this is why volume in the pre-alpha version
+   ;; only goes up to 9.
+   :Digit0 [:attr :Digit0]
+   :Digit1 [:attr :Digit1]
+   :Digit2 [:attr :Digit2]
+   :Digit3 [:attr :Digit3]
+   :Digit4 [:attr :Digit4]
+   :Digit5 [:attr :Digit5]
+   :Digit6 [:attr :Digit6]
+   :Digit7 [:attr :Digit7]
+   :Digit8 [:attr :Digit8]
+   :Digit9 [:attr :Digit9]
+
+   :KeyX      :off
+   :ShiftKeyX :stop
+
+   :Backspace [:macro {:type :simple :value [[:attr nil] (:up-line -motions)]}]
+   :Space     [:position (:down-line -motions)]})
+
+(def keymap
```

Comment:
`keymap-for-mode`?

Reply:
going with `mode-keymaps` https://github.com/cqsd/chipper/commit/86ec97dda246eeffeffbe86c90cad5bf6d5752e0